rig-core/src/embeddings/embeddable.rs

```diff
@@ -0,0 +1,359 @@
+#[derive(Debug, thiserror::Error)]
+pub enum EmbeddableError {
+    #[error("SerdeError: {0}")]
+    SerdeError(#[from] serde_json::Error),
+}
+
+/// Trait for types that can be embedded.
+/// The `embeddable` method returns a OneOrMany<String> which contains strings for which embeddings will be generated by the embeddings builder.
+/// If there is an error generating the list of strings, the method should return an error that implements `std::error::Error`.
+pub trait Embeddable {
+    type Error: std::error::Error;
+
+    fn embeddable(&self) -> Result<OneOrMany<String>, Self::Error>;
+}
+
+/// Struct containing either a single item or a list of items of type T.
+/// If a single item is present, `first` will contain it and `rest` will be empty.
+/// If multiple items are present, `first` will contain the first item and `rest` will contain the rest.
+/// IMPORTANT: this struct cannot be created with an empty vector.
+/// OneOrMany objects can only be created using OneOrMany::one() or OneOrMany::many().
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct OneOrMany<T> {
+    /// First item in the list.
+    first: T,
+    /// Rest of the items in the list.
+    rest: Vec<T>,
+}
+
+impl<T: Clone> OneOrMany<T> {
+    /// Get the first item in the list.
+    pub fn first(&self) -> T {
+        self.first.clone()
+    }
+
+    /// Get the rest of the items in the list (excluding the first one).
+    pub fn rest(&self) -> Vec<T> {
+        self.rest.clone()
+    }
+
+    /// Create a OneOrMany object with a single item of any type.
+    pub fn one(item: T) -> Self {
+        OneOrMany {
+            first: item,
+            rest: vec![],
+        }
+    }
+
+    /// Create a OneOrMany object with a vector of items of any type.
+    pub fn many(items: Vec<T>) -> Self {
+        let mut iter = items.into_iter();
+        OneOrMany {
+            first: match iter.next() {
+                Some(item) => item,
+                None => panic!("Cannot create OneOrMany with an empty vector."),
+            },
+            rest: iter.collect(),
+        }
+    }
+
+    /// Use the Iterator trait on OneOrMany
+    pub fn iter(&self) -> OneOrManyIterator<T> {
+        OneOrManyIterator {
+            one_or_many: self,
+            index: 0,
+        }
+    }
+}
+
+/// Implement Iterator for OneOrMany.
+/// Iterates over all items in both `first` and `rest`.
+/// Borrows the OneOrMany object that is being iterator over.
+pub struct OneOrManyIterator<'a, T> {
+    one_or_many: &'a OneOrMany<T>,
+    index: usize,
+}
+
+impl<'a, T> Iterator for OneOrManyIterator<'a, T> {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut item = None;
+        if self.index == 0 {
+            item = Some(&self.one_or_many.first)
+        } else if self.index - 1 < self.one_or_many.rest.len() {
+            item = Some(&self.one_or_many.rest[self.index - 1]);
+        };
+
+        self.index += 1;
+        item
+    }
+}
```

Comment:
Idk if this is the best option as opposed to simply implementing `IntoIterator` for `OneOrMany` like so:
```rust
impl<T: Clone> IntoIterator for OneOrMany<T> {
    type Item = T;
    type IntoIter = std::iter::Chain<std::iter::Once<T>, std::vec::IntoIter<T>>;

    fn into_iter(self) -> Self::IntoIter {
        std::iter::once(self.first).chain(self.rest)
    }
}
```